core/src/main/java/jenkins/util/VirtualFile.java

```diff
@@ -119,13 +166,62 @@
      */
     public abstract @Nonnull VirtualFile[] list() throws IOException;
 
+    /**
+     * @deprecated use {@link #list(String, String, boolean)} instead
+     */
+    @Deprecated
+    public @Nonnull String[] list(String glob) throws IOException {
+        return list(glob.replace('\\', '/'), null, true).toArray(new String[0]);
+    }
+
     /**
      * Lists recursive files of this directory with pattern matching.
-     * @param glob an Ant-style glob
-     * @return a list of relative names of children (files directly inside or in subdirectories)
+     * <p>The default implementation calls {@link #list()} recursively and applies filtering to the result.
+     * Implementations may wish to override this more efficiently.
+     * @param includes comma-separated Ant-style globs as per {@link Util#createFileSet(File, String, String)} using {@code /} as a path separator;
+     *                 the empty string means <em>no matches</em> (use {@link SelectorUtils#DEEP_TREE_MATCH} if you want to match everything except some excludes)
+     * @param excludes optional excludes in similar format to {@code includes}
+     * @param useDefaultExcludes as per {@link AbstractFileSet#setDefaultexcludes}
+     * @return a list of {@code /}-separated relative names of children (files directly inside or in subdirectories)
      * @throws IOException if this is not a directory, or listing was not possible for some other reason
+     * @since FIXME
      */
-    public abstract @Nonnull String[] list(String glob) throws IOException;
+    public @Nonnull Collection<String> list(@Nonnull String includes, @CheckForNull String excludes, boolean useDefaultExcludes) throws IOException {
+        List<String> r = new ArrayList<>();
+        collectFiles(r, "");
+        List<TokenizedPattern> includePatterns = patterns(includes);
+        List<TokenizedPattern> excludePatterns = patterns(excludes);
+        if (useDefaultExcludes) {
+            for (String patt : DirectoryScanner.getDefaultExcludes()) {
+                excludePatterns.add(new TokenizedPattern(patt.replace('/', File.separatorChar)));
+            }
+        }
+        return r.stream().filter(p -> {
+            TokenizedPath path = new TokenizedPath(p.replace('/', File.separatorChar));
+            return includePatterns.stream().anyMatch(patt -> patt.matchPath(path, true)) && !excludePatterns.stream().anyMatch(patt -> patt.matchPath(path, true));
+        }).collect(Collectors.toSet());
+    }
+    private void collectFiles(Collection<String> names, String prefix) throws IOException {
+        for (VirtualFile child : list()) {
+            if (child.isFile()) {
+                names.add(prefix + child.getName());
+            } else if (child.isDirectory()) {
+                child.collectFiles(names, prefix + child.getName() + "/");
```

Comment:
looking at some implementations `child.getName()` already ends with `/` and this adds an extra one
ie. https://github.com/jenkinsci/compress-artifacts-plugin/blob/ef69f52cfec9edc69443ee11c1435c15febe0329/src/main/java/org/jenkinsci/plugins/compress_artifacts/ZipStorage.java#L126